metals/src/main/scala/scala/meta/internal/metals/codelenses/RunTestCodeLens.scala

```diff
@@ -92,37 +97,72 @@ final class RunTestCodeLens(
       // although hasDebug is already available in BSP capabilities
       // see https://github.com/build-server-protocol/build-server-protocol/pull/161
       // most of the bsp servers such as bloop and sbt might not support it.
-    } yield requestJvmEnvironment(buildTargetId, isJVM).map { _ =>
-      val classes = buildTargetClasses.classesOf(buildTargetId)
+    } yield isNonMetaFileInClasspath(buildTargetId).flatMap {
+      (isNonMeta: Boolean) =>
+        if (isNonMeta) {
+          requestJvmEnvironment(buildTargetId, isJVM).map { _ =>
+            val classes = buildTargetClasses.classesOf(buildTargetId)
 
-      // sbt doesn't declare debugging provider
-      def buildServerCanDebug =
-        connection.isDebuggingProvider || connection.isSbt
-
-      if (connection.isScalaCLI && path.isAmmoniteScript) {
-        scalaCliCodeLenses(
-          textDocument,
-          buildTargetId,
-          classes,
-          distance,
-          buildServerCanDebug,
-          isJVM,
-        )
-      } else
-        codeLenses(
-          textDocument,
-          buildTargetId,
-          classes,
-          distance,
-          path,
-          isJVM,
-        )
+            // sbt doesn't declare debugging provider
+            def buildServerCanDebug =
+              connection.isDebuggingProvider || connection.isSbt
 
+            if (connection.isScalaCLI && path.isAmmoniteScript) {
+              scalaCliCodeLenses(
+                textDocument,
+                buildTargetId,
+                classes,
+                distance,
+                buildServerCanDebug,
+                isJVM,
+              )
+            } else
+              codeLenses(
+                textDocument,
+                buildTargetId,
+                classes,
+                distance,
+                path,
+                isJVM,
+              )
+          }
+        } else Future.successful(Nil)
     }
-
     lenses.getOrElse(Future.successful(Nil))
   }
 
+  /**
+   * Checks if there exist any files outside of the META-INF directory of build target classpath.
+   *
+   * When using Scala 3 Best Effort compilation, the compilation may fail,
+   * but semanticDB and betasty (and thus the META_INF directory) may still be produced.
+   * We want to avoid creating run/test/debug code lens in those situations.
+   */
+  private def isNonMetaFileInClasspath(
```

Comment:
Could we instead just check if the current build target is compiling? `diagnostics.hasCompilationErrors(buildId)` should be faster than looking up filesystem

Reply:
That's true, but there is a case where a project may not produce any errors/diagnostics, but may be compiled based on betasty produced by a failing dependency. In that case we will still not get any artifacts there (outside of betasty/semanticdb), and trying to run will produce classpath errors. 

Though now that I think about it, we could check diagnostics of every dependency of a project to guess if a project was compiled using betasty. I'll try doing that

Reply:
After reviewing my changes just now, I realized that I do the same thing in doctor, not sure why I didn't think to do this here too until now 😅